cpp/src_prims/timeSeries/arima_helpers.h

```diff
@@ -0,0 +1,328 @@
+  DataT coef0 = _param_to_poly<isAr>(param + bid * lags, lags, idx0);
+  DataT coef1 = _param_to_poly<isAr>(sparam + bid * slags, slags, idx1);
+  return isAr ? -coef0 * coef1 : coef0 * coef1;
+}
+
+/**
+ * @brief Prepare data by differencing if needed (simple and/or seasonal)
+ *        and removing a trend if needed
+ *
+ * @note: It is assumed that d + D <= 2. This is enforced on the Python side
+ *
+ * @param[out] d_out       Output. Shape (n_obs - d - D*s, batch_size) (device)
+ * @param[in]  d_in        Input. Shape (n_obs, batch_size) (device)
+ * @param[in]  batch_size  Number of series per batch
+ * @param[in]  n_obs       Number of observations per series
+ * @param[in]  d           Order of simple differences (0, 1 or 2)
+ * @param[in]  D           Order of seasonal differences (0, 1 or 2)
+ * @param[in]  s           Seasonal period if D > 0
+ * @param[in]  stream      CUDA stream
+ * @param[in]  intercept   Whether the model fits an intercept
+ * @param[in]  d_mu        Mu array if intercept > 0
+ *                         Shape (batch_size,) (device)
+ */
+template <typename DataT>
+static void prepare_data(DataT* d_out, const DataT* d_in, int batch_size,
+                         int n_obs, int d, int D, int s, cudaStream_t stream,
+                         int intercept = 0, const DataT* d_mu = nullptr) {
+  // Only one difference (simple or seasonal)
+  if (d + D == 1) {
+    int period = d ? 1 : s;
+    int tpb = (n_obs - period) > 512 ? 256 : 128;  // quick heuristics
+    MLCommon::LinAlg::Batched::
+      batched_diff_kernel<<<batch_size, tpb, 0, stream>>>(d_in, d_out, n_obs,
+                                                          period);
+    CUDA_CHECK(cudaPeekAtLastError());
+  }
+  // Two differences (simple or seasonal or both)
+  else if (d + D == 2) {
+    int period1 = d ? 1 : s;
+    int period2 = d == 2 ? 1 : s;
+    int tpb = (n_obs - period1 - period2) > 512 ? 256 : 128;
+    MLCommon::LinAlg::Batched::
+      batched_second_diff_kernel<<<batch_size, tpb, 0, stream>>>(
+        d_in, d_out, n_obs, period1, period2);
+    CUDA_CHECK(cudaPeekAtLastError());
+  }
+  // If no difference and the pointers are different, copy in to out
+  else if (d + D == 0 && d_in != d_out) {
+    MLCommon::copy(d_out, d_in, n_obs * batch_size, stream);
+  }
+  // Other cases: no difference and the pointers are the same, nothing to do
+
+  // Remove trend in-place
+  if (intercept) {
+    MLCommon::LinAlg::matrixVectorOp(
+      d_out, d_out, d_mu, batch_size, n_obs - d - D * s, false, true,
+      [] __device__(DataT a, DataT b) { return a - b; }, stream);
+  }
+}
+
+/**
+ * @brief Helper function that will read in src0 if the given index is
+ *        negative, src1 otherwise.
+ * @note  This is useful when one array is the logical continuation of
+ *        another and the index is expressed relatively to the second array.
+ */
+template <typename DataT>
+inline __device__ DataT _select_read(const DataT* src0, int size0,
```

Comment:
Please use the `DI` keyword, instead. [Ref](https://github.com/rapidsai/cuml/blob/branch-0.13/cpp/src_prims/cuda_utils.h#L27)